### DISCUSSION TITLE
how should one obtain the decoded parsed payload of a public paseto token?

### DISCUSSION THREAD

**Discussion by iambumblehead** [panva/paseto#30] [Q&A]
I'm trying something like this which returns a string with embedded json followed by non-json "noise".
```javascript
atob( publicpaseto.split('2') );
```
is there a straight-forward way you would recommended to obtain the JSON payload without verification, such as the method shown for JWT tokens at this stackoverflow link? https://stackoverflow.com/questions/38552003/how-to-decode-jwt-token-in-javascript-without-using-a-library#answer-38552302

**Comment by panva** [comment]
I'm sure the project documentation will be useful!

**Reply by iambumblehead** [reply]
I see :) thank you https://github.com/panva/paseto/tree/main/docs#decode

**Reply by iambumblehead** [reply]
```javascript
JSON.parse( Buffer.from( publicpasetoV4.split( '.' )[2], 'base64' ).subarray( 0, -64 ) );
```

**Reply by panva** [reply]
That snippet is going to obviously fail you on different paseto versions.

**Reply by iambumblehead** [reply]
Buffer is not defined in the browser, but this seems to work for v4 tokens
```javascript
 const base64 = publicpasetoV4.split( '.' )[2].replace( /-/g, '+' ).replace( /_/g, '/' );

 JSON.parse( decodeURIComponent( atob( base64 ).slice( 0, -64 )
    .split( '' )
    .map( c => '%' + ( '00' + c.charCodeAt( 0 ).toString( 16 ) ).slice( -2 ) )
    .join( '' ) ) );
```
